src/librustc_resolve/diagnostics.rs

```diff
@@ -964,6 +964,31 @@ impl<'a> Resolver<'a> {
 
         err.emit();
     }
+
+    crate fn make_undeclared_type_suggestion(
+        &mut self,
+        ident: Ident,
+    ) -> (String, Option<Suggestion>) {
+        let parent_scope = &ParentScope::module(self.graph_root);
+        let typo_suggestion = self.early_lookup_typo_candidate(
```

Comment:
This should be either `lookup_typo_candidate` or `early_lookup_typo_candidate` depending on whether `resolve_ident_in_lexical_scope` or `early_resolve_ident_in_lexical_scope` was used above.

Reply:
Hm, do you have a suggestion on how to best propagate this information? I guess I could re-use the conditionals from here:

https://github.com/rust-lang/rust/blob/8040bc983669fcc3fe358dfc9772ea2ecde0ae1b/src/librustc_resolve/lib.rs#L2071

But copy/pasting the conditions will for sure introduce mistakes at some point.

I guess another (naive) approach could be to change `binding` type from `Result<&NameBinding, Determinacy>` to `(binding_source, Result<&NameBinding, Determinacy>)` where `binding_source` contains the required information.

Reply:
You can make a single boolean
```
let use_early_resolution = ribs.is_none() || opt_ns.is_none() || opt_ns == Some(MacroNS);
```
in `fn resolve_path_with_ribs` and pass it as a whole to `make_undeclared_type_suggestion`.

It doesn't matter much, `early_resolve_ident_in_lexical_scope` and `resolve_ident_in_lexical_scope` need to be merged into a single function anyway.